File: comic_editor.cpp
// comic_editor.cpp
#include <Magick++.h>
#include <iostream>
#include <string>
#include <vector>
#include <cstring>

using namespace std;
using namespace Magick;

void drawCircle(Image& img, int x, int y, int w, int h, const Color& fill, const Color& border) {
    // Simple circle using flood fill (simplified)
    // For a real implementation, use drawing functions
    img.strokeColor(border);
    img.fillColor(fill);
    img.draw(DrawableCircle(x + w/2, y + h/2, w/2, 0));
}

void drawSquare(Image& img, int x, int y, int w, int h, const Color& fill, const Color& border) {
    img.strokeColor(border);
    img.fillColor(fill);
    img.draw(DrawableRectangle(x, y, x+w, y+h));
}

void drawCloud(Image& img, int x, int y, int w, int h, const Color& fill, const Color& border) {
    // Simplified cloud with overlapping circles
    drawCircle(img, x + w/3, y + h/3, w/2, h/2, fill, border);
    drawCircle(img, x + w/6, y + h/4, w/2, h/2, fill, border);
    drawCircle(img, x + w/2, y + h/4, w/2, h/2, fill, border);
}

int main(int argc, char* argv[]) {
    InitializeMagick(nullptr);

    string input, text, shape = "circle", bubbleColor = "#FFFFFF", borderColor = "#000000", textColor = "#000000", output = "comic.png";
    int x = 50, y = 50, w = 200, h = 100, fontSize = 16;

    for (int i = 1; i < argc; ++i) {
        string arg = argv[i];
        if (arg == "--input" && i+1 < argc) input = argv[++i];
        else if (arg == "--text" && i+1 < argc) text = argv[++i];
        else if (arg == "--x" && i+1 < argc) x = stoi(argv[++i]);
        else if (arg == "--y" && i+1 < argc) y = stoi(argv[++i]);
        else if (arg == "--shape" && i+1 < argc) shape = argv[++i];
        else if (arg == "--width" && i+1 < argc) w = stoi(argv[++i]);
        else if (arg == "--height" && i+1 < argc) h = stoi(argv[++i]);
        else if (arg == "--bubble-color" && i+1 < argc) bubbleColor = argv[++i];
        else if (arg == "--border-color" && i+1 < argc) borderColor = argv[++i];
        else if (arg == "--text-color" && i+1 < argc) textColor = argv[++i];
        else if (arg == "--font-size" && i+1 < argc) fontSize = stoi(argv[++i]);
        else if (arg == "--output" && i+1 < argc) output = argv[++i];
    }

    if (input.empty() || text.empty()) {
        cerr << "Error: --input and --text are required" << endl;
        return 1;
    }

    try {
        Image img(input);

        Color fill(bubbleColor);
        Color border(borderColor);

        if (shape == "circle") {
            drawCircle(img, x, y, w, h, fill, border);
        } else if (shape == "square") {
            drawSquare(img, x, y, w, h, fill, border);
        } else if (shape == "cloud") {
            drawCloud(img, x, y, w, h, fill, border);
        }

        // Draw text (simplified)
        img.fontPointsize(fontSize);
        img.fillColor(Color(textColor));
        img.annotate(text, Geometry(w, h), GravityCenter);

        img.write(output);
        cout << "Comic saved to " << output << endl;
    } catch (const exception& e) {
        cerr << "Error: " << e.what() << endl;
        return 1;
    }

    return 0;
}

File: comic_editor.js
#!/usr/bin/env node
// comic_editor.js
const { program } = require('commander');
const sharp = require('sharp');
const fs = require('fs');
const chalk = require('chalk');

class ComicEditor {
    constructor(options) {
        this.inputPath = options.input;
        this.text = options.text;
        this.x = options.x || 50;
        this.y = options.y || 50;
        this.shape = options.shape || 'circle';
        this.width = options.width || 200;
        this.height = options.height || 100;
        this.bubbleColor = options.bubbleColor || '#FFFFFF';
        this.borderColor = options.borderColor || '#000000';
        this.textColor = options.textColor || '#000000';
        this.fontSize = options.fontSize || 16;
        this.output = options.output || 'comic.png';
    }

    async run() {
        try {
            // Создаём SVG для облачка
            let svg = '';
            const x = this.x, y = this.y, w = this.width, h = this.height;
            const shape = this.shape;
            const bc = this.bubbleColor;
            const br = this.borderColor;

            if (shape === 'circle') {
                svg = `<ellipse cx="${x + w/2}" cy="${y + h/2}" rx="${w/2}" ry="${h/2}" fill="${bc}" stroke="${br}" stroke-width="2"/>`;
            } else if (shape === 'square') {
                svg = `<rect x="${x}" y="${y}" width="${w}" height="${h}" fill="${bc}" stroke="${br}" stroke-width="2"/>`;
            } else if (shape === 'cloud') {
                svg = `<circle cx="${x + w/3}" cy="${y + h/3}" r="${w/3}" fill="${bc}" stroke="${br}" stroke-width="2"/>
                       <circle cx="${x + w*2/3}" cy="${y + h/3}" r="${w/3}" fill="${bc}" stroke="${br}" stroke-width="2"/>
                       <circle cx="${x + w/2}" cy="${y + h*2/3}" r="${w/3}" fill="${bc}" stroke="${br}" stroke-width="2"/>`;
            }

            // Текст
            const textSvg = `<text x="${x + w/2}" y="${y + h/2}" text-anchor="middle" dominant-baseline="central"
                             font-family="Arial" font-size="${this.fontSize}" fill="${this.textColor}">${this.text}</text>`;
            const fullSvg = `<svg width="${this.width}" height="${this.height}">${svg}${textSvg}</svg>`;

            // Накладываем на изображение
            const overlay = Buffer.from(fullSvg);
            await sharp(this.inputPath)
                .composite([{ input: overlay, top: 0, left: 0 }])
                .toFile(this.output);
            console.log(chalk.green(`Comic saved to ${this.output}`));
        } catch (err) {
            console.error(chalk.red(`Error: ${err.message}`));
            process.exit(1);
        }
    }
}

program
    .requiredOption('--input <file>', 'Input image file')
    .requiredOption('--text <text>', 'Text in the bubble')
    .option('--x <number>', 'X coordinate', parseInt, 50)
    .option('--y <number>', 'Y coordinate', parseInt, 50)
    .option('--shape <type>', 'circle, square, cloud', 'circle')
    .option('--width <number>', 'Bubble width', parseInt, 200)
    .option('--height <number>', 'Bubble height', parseInt, 100)
    .option('--bubble-color <color>', 'Bubble color', '#FFFFFF')
    .option('--border-color <color>', 'Border color', '#000000')
    .option('--text-color <color>', 'Text color', '#000000')
    .option('--font-size <number>', 'Font size', parseInt, 16)
    .option('--output <file>', 'Output file', 'comic.png')
    .parse(process.argv);

const opts = program.opts();
const editor = new ComicEditor(opts);
editor.run();
